streams-tests/jvm/src/test/scala/zio/stream/StreamInterruptWhenSpec.scala

```diff
@@ -10,31 +10,61 @@ object StreamInterruptWhenSpec extends ZIOBaseSpec {
-          started     <- Promise.make[Nothing, Unit]
-          fiber <- ZStream
-                    .fromEffect((started.succeed(()) *> latch.await).onInterrupt(interrupted.set(true)))
-                    .interruptWhen(halt)
-                    .runDrain
-                    .fork
-          _      <- started.await *> halt.succeed(())
-          _      <- fiber.await
-          result <- interrupted.get
-        } yield assert(result)(isTrue)
-      },
-      testM("propagates errors") {
-        for {
-          halt <- Promise.make[String, Nothing]
-          _    <- halt.fail("Fail")
-          result <- ZStream(1)
-                     .haltWhen(halt)
-                     .runDrain
-                     .either
-        } yield assert(result)(isLeft(equalTo("Fail")))
-      } @@ zioTag(errors)
-    ) @@ zioTag(interruption)
+      suite("ZStream.interruptWhen(Promise)")(
+        testM("interrupts the current element") {
+          for {
+            interrupted <- Ref.make(false)
+            latch       <- Promise.make[Nothing, Unit]
+            halt        <- Promise.make[Nothing, Unit]
+            started     <- Promise.make[Nothing, Unit]
+            fiber <- ZStream
+                      .fromEffect((started.succeed(()) *> latch.await).onInterrupt(interrupted.set(true)))
+                      .interruptWhen(halt)
+                      .runDrain
+                      .fork
+            _      <- started.await *> halt.succeed(())
+            _      <- fiber.await
+            result <- interrupted.get
+          } yield assert(result)(isTrue)
+        },
+        testM("propagates errors") {
+          for {
+            halt <- Promise.make[String, Nothing]
+            _    <- halt.fail("Fail")
+            result <- ZStream(1)
+                       .haltWhen(halt)
+                       .runDrain
+                       .either
+          } yield assert(result)(isLeft(equalTo("Fail")))
+        } @@ zioTag(errors)
+      ) @@ zioTag(interruption),
+      suite("ZStream.interruptWhen(IO)")(
+        testM("interrupts the current element") {
+          for {
+            interrupted <- Ref.make(false)
+            latch       <- Promise.make[Nothing, Unit]
+            halt        <- Promise.make[Nothing, Unit]
+            started     <- Promise.make[Nothing, Unit]
+            fiber <- ZStream
+                      .fromEffect((started.succeed(()) *> latch.await).onInterrupt(interrupted.set(true)))
+                      .interruptWhen(halt.await)
+                      .runDrain
+                      .fork
+            _      <- started.await *> halt.succeed(())
+            _      <- fiber.await
+            result <- interrupted.get
+          } yield assert(result)(isTrue)
+        },
+        testM("propagates errors") {
+          for {
+            halt <- Promise.make[String, Nothing]
+            _    <- halt.fail("Fail")
+            result <- ZStream.fromIterable(scala.Stream.from(0))
```

Comment:
@iravid just want to call this out – I had to change this from `ZStream(1)` because if the ZStream is already successful and is small, then there is no chance to observe the completion of `halt.await` (even though it's already done) and the resulting stream doesn't fail like it should. I'm not sure if that's an issue or not, in real life (what does it mean to halt/interrupt an already successful stream?)